pkg/credentialprovider/apis/criauth/v1alpha1/types.go

```diff
@@ -0,0 +1,118 @@
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CRIAuthPluginConfig is the configuration containing information about
+// each CRI auth plugin. Kubelet reads this configuration from disk and enables
+// each plugin as specified by the CRIAuthPlugin type.
+type CRIAuthPluginConfig struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Plugins is a list of CRI auth plugins that will be enabled by the kubelet
+	Plugins []CRIAuthPlugin `json:"plugins"`
+}
+
+// CRIAuthPlugin is used by the kubelet container runtime to match the
+// image property string (from the container spec) with exec-based credential provider
+// plugins that provide container registry credentials.
+type CRIAuthPlugin struct {
+	// name is the name of the credential plugin.  It must match the name of the
+	// executable as seen by the kubelet.
+	Name string `json:"name"`
+
+	// imageMatchRefs contains information about the matching image URLs this plugin should
+	// handle along with the cache duration of credentials for any matching images.
+	// Credentials are cached independently per imageMatchRef.
+	ImageMatchRefs []ImageMatchRef `json:"imageMatchRefs"`
+
+	// ExtraArgs specifies arguments to be passed to the plugin via environment variables.
+	// +optional
+	ExtraArgs []PluginArg `json:"extraArgs,omitempty"`
+}
+
+// ImageMatchRef holds information for what image URLs a plugin should handle.
+// The plugin will be called by a kubelet if an image requiring credentials matches
+// the information in imageMatchRef. It also indicates how long the auth plugin
+// should cache credentials in-memory.
+type ImageMatchRef struct {
+	// matchURL is a list of strings used to match against the image property
+	// (sometimes called "registry path") to determine which images to provide
+	// credentials for.  If one of the strings matches the image property, then the
+	// CRIAuthPlugin will be used by kubelet to provide credentials
+	// for the image pull.
+	//
+	// The image property of a container supports the same syntax as the docker
+	// command does, including private registries and tags. A registry path is
+	// similar to a URL, but does not contain a protocol specifier (https://).
+	//
+	// Each matchURL string is a pattern which can optionally contain
+	// a port and a path, similar to the image spec.  Globs can be used in the
+	// hostname (but not the port or the path).
+	//
+	// Globs are supported as subdomains (*.k8s.io) or (k8s.*.io), and
+	// top-level-domains (k8s.*).  Matching partial subdomains is also supported
+	// (app*.k8s.io).  Each glob can only match a single subdomain segment, so
+	// *.io does not match *.k8s.io.
+	//
+	// The image property matches when it has the same number of parts as the
+	// ImageMatcher string, and each part matches.  Additionally the path of
+	// ImageMatcher must be a prefix of the target URL. If the ImageMatcher
+	// contains a port, then the port must match as well.
```

Comment:
how does a plugin indicate it provides credentials for all images (e.g. should be used as a backstop if there's not an imagePullSecret that provides credentials for a given image)?

Reply:
As discussed last week, this is out of scope for the first iteration of this feature since we are trying to acheive feature parity with the existing in-tree providers first.